### PR TITLE
SIL: add the borrowed-from instruction

### PR DESCRIPTION
**Pull request by eeckstein**
The borrowed-from instruction declares from which enclosing values a guaranteed phi argument is borrowed from. For example:

```
    %2 = begin_borrow %1 : $T
    br bb1(%2 : $T, %1 : $T)
  bb1(%3: @guaranteed $T, %4 : @owned $T):
    %5 = borrowed %3 : $T from (%4 : $T)
```

An enclosing value is either a dominating borrow introducer of the borrowed operand (``%3``) or an adjacent phi-argument in the same block (``%4``).

So far, this information was computed dynamically by ownership utilities (e.g. `gatherEnclosingValues`). With the borrowed-from instructions those utilities get dramatically simpler.

The borrowed-from instruction is similar in spirit to phi-nodes in plain SSA: it "caches" information from predecessor blocks so that optimization passes can easily access this information. If an optimization pass introduces a new guaranteed phi argument it can use the `BorrowedFromUpdater` to generate the required borrowed-from instructions. This is similar to `SSAUpdater` which inserts required phi-nodes.

Beside simplifying the ownership utilities, it's also useful that the adjacent phi relations are now explicitly visible in SIL.

Another nice thing is that this change is (almost) a pure addition to SIL. This means that the existing C++ ownership utilities still work - with and without inserted borrowed-from instructions. The only exception is that they have to look through borrowed-from instructions in some places.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test

File: docs/SIL.rst [hunk 1]
```diff
@@ -4615,6 +4615,38 @@ We require that ``%1`` and ``%0`` have the same type ignoring SILValueCategory.
...
+operand in all predecessor blocks.
+
+The borrowed operand (``%2``) must be a guaranteed phi argument and is
+forwarded to the instruction result.
+
+The list of enclosing values (operands after ``from``) can be empty if the
+borrowed operand stems from a borrow introducer with no enclosing value, e.g.
+a ``load_borrow``.
+
+Guaranteed phi arguments must not have other users than borrowed-from
```
Comment (atrick) [inline docs/SIL.rst:4645]: > Guaranteed phi arguments must not have other users than borrowed-from

This is not true for reborrows. Maybe it's true for guaranteed forwarding phis.

File: docs/SIL.rst [hunk 2]
```diff
@@ -4615,6 +4615,38 @@ We require that ``%1`` and ``%0`` have the same type ignoring SILValueCategory.
...
+     // %3 has type $T and guaranteed ownership
+
+Declares from which enclosing values a guaranteed phi argument is borrowed
+from.
+An enclosing value is either a dominating borrow introducer (``%0``) of the
+borrowed operand (``%2``) or an adjacent phi-argument in the same block
+(``%1``).
+In case of an adjacent phi, all incoming values of the adjacent phi must be
+borrow introducers for the corresponding incoming value of the borrowed
+operand in all predecessor blocks.
```
Comment (meg-gupta) [inline docs/SIL.rst:4636]: While working on guaranteed phis, I think I encountered some cases where enclosing value could be adjacent or dominating. Does the borrowed from insertion pass handle this ?

Comment (eeckstein) [reply docs/SIL.rst:4636]: I'm not sure what you mean exactly. Yes, the enclosing values (= the operands after `from`) can be an adjacent phis or dominating borrow introducers. 
This is computed iteratively (until convergence) from incoming values in predcessors

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Pull request merged**